src/ol/renderer/canvas/VectorLayer.js

```diff
@@ -328,7 +338,7 @@ class CanvasVectorLayerRenderer extends CanvasLayerRenderer {
    */
   prepareFrame(frameState) {
     const vectorLayer = this.getLayer();
-    const vectorSource = vectorLayer.getSource();
+    const vectorSource = vectorLayer.getSource() || emptySource;
```

Comment:
What happens if we return early here instead of proceeding with an empty source?

Reply:
Returning false seems to have no adverse effect (it simply stops the render).  I had been concerned that if the source was cleared using `setSource(null)` the renderer's properties would not get reset but without any rendering that should not be a problem, and it also avoids a similar error in the renderFrame function if interacting when `setSource(null)` is called.